### PR TITLE
feat: Several output additions and improvements.

### PR DESCRIPTION
**Pull request by jetuk**
- Fixed an issue where metric sets were not finalised at the end of a model run.
- Added a new Memory output and associated recorder. This recorder stores data in memory.
  - TODO: improvement the error handling in the aggregation function.
- Add new functions to Recorder API to allow accessing an aggregated metric.
  - Additional work is needed to give access to different levels of aggregation.
  - Currently only the new Memory recorder supports this API.
  - The order of the aggregation operations probably needs looking at.
- Changed serialised format of OutputMetric. This will allow it it eventually contain additional information.
- Changed serialised format of MetricAggFunc. Again, this will allow definition of functions with additional arguments.
- Added CountNonZero aggregation function.

### CODE REVIEW

**Review by Batch21 — APPROVED** [review]
A few comments but the updates look good.

Probably worth adding some of the TODOs you've noted as issues when this is merged.

File: pywr-core/src/recorders/aggregator.rs [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
     Annual,
+    Rolling { window: NonZeroUsize },
 }
 
 impl AggregationFrequency {
     fn is_date_in_period(&self, period_start: &Date, date: &Date) -> bool {
         match self {
             Self::Monthly => (period_start.year() == date.year()) && (period_start.month() == date.month()),
             Self::Annual => period_start.year() == date.year(),
+            Self::Rolling { window } => {
```
[the remark this answers is not in the document]
Comment (jetuk) [reply pywr-core/src/recorders/aggregator.rs:16]: I should probably pull this out as a separate PR.

[the remark this answers is not in the document]
Comment (jetuk) [reply pywr-core/src/recorders/aggregator.rs]: I've just renamed this to `Days` for now. I think a proper rolling aggregator will not fit in the periodic framework. Separately we'll need a different aggregator for rolling windows, and then maybe another one for events. Somehow these will need to fit into the same API.

File: pywr-core/src/recorders/memory.rs [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+                aggregation.apply_time_func(&ts)
+            })
+            .collect::<Result<_, _>>()?;
+
+        aggregation.apply_scenario_func(&scenario_data)
+    }
+
+    /// Aggregate over the saved data to a single value using the provided aggregation functions.
+    ///
+    /// This method will first aggregation over the metrics, then over time, and finally over the scenarios.
```
Comment (Batch21) [inline pywr-core/src/recorders/memory.rs:170]: comment needs correcting?

Comment (jetuk) [reply pywr-core/src/recorders/memory.rs]: Done

**Comment by jetuk** [comment]
@Batch21 were you happy with this to be merged?

**Comment by Batch21** [comment]
> @Batch21 were you happy with this to be merged?

Yeah, go for it

**Pull request merged**